### PR TITLE
google-chrome: 127.0.6533.89 -> 127.0.6533.100, add darwin support

### PR DESCRIPTION
**Pull request by 4evy**
Closes: https://github.com/NixOS/nixpkgs/pull/162436
Closes: https://github.com/NixOS/nixpkgs/pull/162467

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by JohnRTitor — COMMENTED** [review]
Linux could also use `stdenvNoCC` I believe, after all why would it need a C compiler? @jnsgruk what do you think?

File: pkgs/by-name/go/google-chrome/package.nix
Comment (superherointj) [inline pkgs/by-name/go/google-chrome/package.nix:110]: I don't know if you realize. But items were ordered alphabetically. I'd prefer if you kept them alphabetically ordered.

File: pkgs/by-name/go/google-chrome/package.nix
Comment (superherointj) [inline pkgs/by-name/go/google-chrome/package.nix:192]: Hook is missing

File: pkgs/by-name/go/google-chrome/package.nix
Comment (superherointj) [inline pkgs/by-name/go/google-chrome/package.nix:8]: Why are you changing this?

File: pkgs/by-name/go/google-chrome/package.nix
Comment (superherointj) [inline pkgs/by-name/go/google-chrome/package.nix:214]: `substituteInPlace`'s `--replace` flag is deprecated.

File: pkgs/by-name/go/google-chrome/package.nix
Comment (superherointj) [inline pkgs/by-name/go/google-chrome/package.nix:179]: > strictDeps = true;

It's a good habit to keep `strictDeps` on, to help separate what belongs to `nativeBuildInputs` and `buildInputs`.

**Comment by DontEatOreo** [comment]
Thanks you for the review

File: pkgs/by-name/go/google-chrome/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/go/google-chrome/package.nix:282]: Uhhh

Comment (DontEatOreo) [reply pkgs/by-name/go/google-chrome/package.nix:282]: I can't actually say for sure because querying old versions on the XML POST request always gives you the latest version, unless I'm doing it wrong... So I would imagine it changes each time

P.S: I notice something interesting tho both Windows x64 and macOS arm64 have almost same hash?

```console
# Windows x32
oct5rowwm5fvvxipazq2wlma2u

# Windows x64
acljsqmj242ipgr5z4ha6ymv5ita

# macOS
aclo5k72baz7erlbweyjruguwh2a
```

Comment (JohnRTitor) [reply pkgs/by-name/go/google-chrome/package.nix:282]: Could you move the update script to a seperate update.sh (for both Linux and Darwin)? The derivation is way too large as it is.

File: pkgs/by-name/go/google-chrome/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/go/google-chrome/package.nix:192]:
```suggestion
  buildInputs = [
    # needed for XDG_ICON_DIRS
    adwaita-icon-theme
    glib
    gtk3
    gtk4
    # needed for GSETTINGS_SCHEMAS_PATH
    gsettings-desktop-schemas
  ];
```

Please keep it the same as they were!

File: pkgs/by-name/go/google-chrome/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/go/google-chrome/package.nix:198]: Good job on adding the hooks, why didn't I fix that sooner.

**Review by JohnRTitor — COMMENTED** [review]
Built successfully on nix-community darwin builder.

```log
johnrtitor@darwin01 ~ % nix build "github:nixos/nixpkgs/refs/pull/332229/head#google-chrome" --print-out-paths --impure
/nix/store/9wr6004ajfaf1jlwv7b6i6673lagvjcm-google-chrome-127.0.6533.89
johnrtitor@darwin01 ~ % ls /nix/store/9wr6004ajfaf1jlwv7b6i6673lagvjcm-google-chrome-127.0.6533.89                     
Applications bin
johnrtitor@darwin01 ~ % cd /nix/store/9wr6004ajfaf1jlwv7b6i6673lagvjcm-google-chrome-127.0.6533.89                   
johnrtitor@darwin01 ..673lagvjcm-google-chrome-127.0.6533.89 % ls Applications
Google Chrome.app
johnrtitor@darwin01 ..673lagvjcm-google-chrome-127.0.6533.89 % ls bin
google-chrome-stable
```

File: pkgs/by-name/go/google-chrome/package.nix
Comment (JohnRTitor) [reply pkgs/by-name/go/google-chrome/package.nix:282]: and would be nice if the darwin package could be updated from linux from using the same update script (obviously we can't run them though). This will make sure it stays updated by ryantm-r (which only runs on x64-linux AFAIK) if we run behind the updates.

And if you want to maintain this, do add yourself to maintainers list.

Comment (DontEatOreo) [reply pkgs/by-name/go/google-chrome/package.nix:282]: > Could you move the update script to a separate update.sh (for both Linux and Darwin)? The derivation is way too large as it is.

Sure! But could you give me some "good" examples of ./update.sh scripts? I was looking at the manual, but it only mentions it was possible without any examples.

I was looking at the repo for update.sh files, but each one is pretty different

> And if you want to maintain this, do add yourself to meta.maintainers.

I don't use Chrome at all, so I won't really be a good maintainer for it :/

Comment (JohnRTitor) [reply pkgs/by-name/go/google-chrome/package.nix:282]: > I don't use Chrome at all, so I won't really be a good maintainer for it :/

K, as long as running

```
nix-shell maintainers/scripts/update.nix --argstr package google-chrome
```
updates both linux and darwin packages I have no issues with it having no darwin maintainers.

I don't think this should be any much of an effort at all, given we are just unpacking and installing the dmg file. I don't have a darwin system so won't be able to test if it runs properly though.

Comment (JohnRTitor) [reply pkgs/by-name/go/google-chrome/package.nix:282]: > Sure! But could you give me some "good" examples of ./update.sh scripts? I was looking at the manual, but it only mentions it was possible without any examples.

You could move the linux part to `update_linux ()` bash function. The same goes for `update_darwin ()`. The main part should call the both functions.

https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/editors/android-studio/update.sh

This could be helpful.

File: pkgs/by-name/go/google-chrome/package.nix
Comment (JohnRTitor) [inline pkgs/by-name/go/google-chrome/package.nix:301]: So this dmg file has both libs for aarch64 and x64 darwin? Interesting.

Comment (DontEatOreo) [reply pkgs/by-name/go/google-chrome/package.nix:301]: Yea, some apps bundle both x86 and arm64 support into single "app"

**Review by JohnRTitor — COMMENTED** [review]
I was testing the update script on my system. It was able to successfully update the linux version, but darwin got stuck due to missing `xmllint`.

```log
Running update for:
 - google-chrome-124.0.6533.88: UPDATING ...
 - google-chrome-124.0.6533.88: ERROR

--- SHOWING ERROR LOG FOR google-chrome-124.0.6533.88 ----------------------

warning: The old format conversion sub commands of `nix hash` were deprecated in favor of `nix hash convert`.
/home/masum/Dev-Environment/nixpkgs/pkgs/by-name/go/google-chrome/update.sh: line 68: xmllint: command not found
```

**Comment by JohnRTitor** [comment]
CC @jnsgruk for testing.

**Comment by jnsgruk** [comment]
So the update-script seems fine to me now:

```
❯ nix-shell maintainers/scripts/update.nix --argstr package google-chrome
this derivation will be built:
  /nix/store/gfsc1r8h09psmgp6m4h5m23dczzh5412-packages.json.drv
these 7 paths will be fetched (8.44 MiB download, 41.84 MiB unpacked):
  /nix/store/p5441mp6wwpdh65qamaixvd092cwr45h-bash-interactive-5.2p26
  /nix/store/dbcw19dshdwnxdv5q2g6wldj6syyvq7l-glibc-2.39-52
  /nix/store/psfazaip7ywf3vzxfd6ngjn9zwmla7gl-libidn2-2.3.7
  /nix/store/n07l5k5kwxd7954wl7xk6v0qwxz536m3-libunistring-1.1
  /nix/store/7ka1d6qli0bci8d1k9b7zw0jrd2s2q1k-ncurses-6.4
  /nix/store/va84gp6s9mk8cyv2k8xmfsai2b33zx7v-readline-8.2p10
  /nix/store/nda0h04bakn2damsd06vkscwi5ds4qjd-xgcc-13.2.0-libgcc
copying path '/nix/store/nda0h04bakn2damsd06vkscwi5ds4qjd-xgcc-13.2.0-libgcc' from 'https://cache.nixos.org'...
building '/nix/store/gfsc1r8h09psmgp6m4h5m23dczzh5412-packages.json.drv'...
copying path '/nix/store/n07l5k5kwxd7954wl7xk6v0qwxz536m3-libunistring-1.1' from 'https://cache.nixos.org'...
sourcing setup hook '/nix/store/dv5vgsw8naxnkcc88x78vprbnn1pp44y-patchelf-0.15.0/nix-support/setup-hook'
sourcing setup hook '/nix/store/i4iynx9axbq23sd0gyrc5wdb46zz6z8l-update-autotools-gnu-config-scripts-hook/nix-support/setup-hook'
sourcing setup hook '/nix/store/h9lc1dpi14z7is86ffhl3ld569138595-audit-tmpdir.sh'
sourcing setup hook '/nix/store/m54bmrhj6fqz8nds5zcj97w9s9bckc9v-compress-man-pages.sh'
sourcing setup hook '/nix/store/wgrbkkaldkrlrni33ccvm3b6vbxzb656-make-symlinks-relative.sh'
sourcing setup hook '/nix/store/5yzw0vhkyszf2d179m0qfkgxmp5wjjx4-move-docs.sh'
sourcing setup hook '/nix/store/fyaryjvghbkpfnsyw97hb3lyb37s1pd6-move-lib64.sh'
sourcing setup hook '/nix/store/kd4xwxjpjxi71jkm6ka0np72if9rm3y0-move-sbin.sh'
sourcing setup hook '/nix/store/pag6l61paj1dc9sv15l7bm5c17xn5kyk-move-systemd-user-units.sh'
sourcing setup hook '/nix/store/jivxp510zxakaaic7qkrb7v1dd2rdbw9-multiple-outputs.sh'
sourcing setup hook '/nix/store/ilaf1w22bxi6jsi45alhmvvdgy4ly3zs-patch-shebangs.sh'
sourcing setup hook '/nix/store/cickvswrvann041nqxb0rxilc46svw1n-prune-libtool-files.sh'
sourcing setup hook '/nix/store/xyff06pkhki3qy1ls77w10s0v79c9il0-reproducible-builds.sh'
sourcing setup hook '/nix/store/aazf105snicrlvyzzbdj85sx4179rpfp-set-source-date-epoch-to-latest.sh'
sourcing setup hook '/nix/store/gps9qrh99j7g02840wv5x78ykmz30byp-strip.sh'
copying path '/nix/store/psfazaip7ywf3vzxfd6ngjn9zwmla7gl-libidn2-2.3.7' from 'https://cache.nixos.org'...
copying path '/nix/store/dbcw19dshdwnxdv5q2g6wldj6syyvq7l-glibc-2.39-52' from 'https://cache.nixos.org'...
copying path '/nix/store/7ka1d6qli0bci8d1k9b7zw0jrd2s2q1k-ncurses-6.4' from 'https://cache.nixos.org'...
copying path '/nix/store/va84gp6s9mk8cyv2k8xmfsai2b33zx7v-readline-8.2p10' from 'https://cache.nixos.org'...
copying path '/nix/store/p5441mp6wwpdh65qamaixvd092cwr45h-bash-interactive-5.2p26' from 'https://cache.nixos.org'...

Going to be running update for following packages:
 - google-chrome-127.0.6533.88

Press Enter key to continue...

Running update for:
 - google-chrome-127.0.6533.88: UPDATING ...
 - google-chrome-127.0.6533.88: DONE.

Packages updated!
```

But when I build the package and try to run the result on Linux I get the following:

```
 ~/code/nixpkgs  add-darwin-support-google-chrome                                                                                                                 venv 
❯ nix build .#legacyPackages.x86_64-linux.google-chrome --impure

 ~/code/nixpkgs  add-darwin-support-google-chrome                                                                                                                 venv 
❯ ./result/bin/google-chrome-stable
./result/bin/google-chrome-stable: line 1042: /nix/store/1vgw377741i15sq8z97shln7k62awwks-google-chrome-127.0.6533.88/share/google/chrome/google-chrome: cannot execute: required file not found
```

**Comment by JohnRTitor** [comment]
> So the update-script seems fine to me now:
It won't actually run all the steps unless there is actually a update available. 

Here's how I reproduced the error:

Set the darwin and Linux chrome version to a older one, like `124.140.8.120`, tinkered the hash for a bit. 

Then I ran the update script.

**Comment by DontEatOreo** [comment]
It works on my side?

https://github.com/user-attachments/assets/adc11381-243a-408a-9a55-6709ae1f3fd1

**Comment by jnsgruk** [comment]
The more concerning failure to me was not being able to run `result/bin/google-chrome` on my machine...

**Comment by DontEatOreo** [comment]
> The more concerning failure to me was not being able to run `result/bin/google-chrome` on my machine...

I'm not sure why that's the case, I haven't remove any deps or changed the installScript...

Do you guys have any idea on what could be causing it? @JohnRTitor @jnsgruk

**Comment by jnsgruk** [comment]
Sorry I ran out of time to get to this later in the week. I'll try to get a better look next week.

**Review by JohnRTitor — APPROVED** [review]
I took a look and linux version works for me.

```log
$  nom build .#google-chrome --impure
warning: Git tree '/home/masum/.cache/nixpkgs-review/pr-332229/nixpkgs' is dirty
Finished at 00:17:18 after 4s

❯  ./result/bin/google-chrome-stable --version
Google Chrome 127.0.6533.88 
```

image

image

Just gotta leave the approval here, though it would be best to wait for jnsgruk's approval.

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 332229` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>google-chrome</li>
    <li>netflix</li>
  </ul>
</details>

**Comment by jnsgruk** [comment]
Hmm, I'm not sure what to do here - I still cannot make this work:

```
 ~/code/nixpkgs  add-darwin-support-google-chrome
❯ nix build .#google-chrome --impure --rebuild

 ~/code/nixpkgs  add-darwin-support-google-chrome                                                                                                ❌127
❯ ./result/bin/google-chrome-stable
./result/bin/google-chrome-stable: line 1042: /nix/store/1vgw377741i15sq8z97shln7k62awwks-google-chrome-127.0.6533.88/share/google/chrome/google-chrome: cannot execute: required file not found
```

I've tried removing the result and rebuilding, but I always end up with the same error :/

**Comment by JohnRTitor** [comment]
Could you try rebasing the branch on top of master? or you could just run it from `nixpkgs-review`.

**Comment by jnsgruk** [comment]
Aha! I've found the problem I think... that file `/nix/store/1vgw377741i15sq8z97shln7k62awwks-google-chrome-127.0.6533.88/share/google/chrome/google-chrome` is a bash script that starts `#!/bin/bash`, rather than `#!/usr/bin/env bash`.

It's a wrapper... and if I run `/nix/store/1vgw377741i15sq8z97shln7k62awwks-google-chrome-127.0.6533.88/share/google/chrome/chrome` the browser starts fine.

@DontEatOreo/@JohnRTitor could you take a look at why that's not being patched (anymore?)? (I can't get much more time on this today, but I could look next week)

**Comment by jnsgruk** [comment]
> Could you try rebasing the branch on top of master? or you could just run it from `nixpkgs-review`.

I've rebased, but I get the same issue. I've pushed the branch anyway so it's up to date

**Comment by JohnRTitor** [comment]
Oh, I have `services.envfs` enabled, which is why I could not catch this bug.

(envfs, provides a `/bin/bash`)

**Comment by JohnRTitor** [comment]
Looks like `strictDeps = true` was causing the error.

It was not in the previous revision, I have explicitly disabled it (though it is disabled by default, anyway). @jnsgruk this should work correctly now.

**Review by jnsgruk — APPROVED** [review]

**Comment by jnsgruk** [comment]
That got it - thank you @JohnRTitor!

I don't think we should be backporting this - given that it's not security related. The diff is quite large, though, so I suspect we'll need to be careful backporting upcoming security releases, and possibly do so manually until the next release.

**Comment by JohnRTitor** [comment]
It's definitely not a security fix but rather can be considered a package addition, which is a valid reason for backport. The diff is indeed large, but the `linux` section is more or less the same (ignoring the indentation changes).

If we choose to not backport this, it will be harder to backport version upgrades in future. The changes for the `linux` section is very insignificant.

**Comment by jnsgruk** [comment]
> It's definitely not a security fix but rather can be considered a package addition, which is a valid reason for backport. The diff is indeed large, but the `linux` section is more or less the same (ignoring the indentation changes).
> 
> 
> 
> If we choose to not backport this, it will be harder to backport version upgrades in future. The changes for the `linux` section is very insignificant. 

I agree. Alright, on the grounds that the Linux side is mostly unchanged, let's backport it - though I'm expecting some manual changes required (as they were in the last)

**Comment by JohnRTitor** [comment]
I think there is an update available. @jnsgruk could you checkout this branch and push the update using the update script? Let's not create another PR for that :)

**Comment by jnsgruk** [comment]
> I think there is an update available. @jnsgruk could you checkout this branch and push the update using the update script? Let's not create another PR for that :)

Done, though I still got that weird `libxml2` issue and had to run it in a shell with `libxml2` explicitly stated

**Comment by JohnRTitor** [comment]
Exactly, I wonder why. This will likely break auto updaters like ryantm bot. @DontEatOreo could you investigate?

At worst we might need to wrap the update script with `makeBinPath` or something.

**Comment by jnsgruk** [comment]
> Exactly, I wonder why. This will likely break auto updaters like ryantm bot. @DontEatOreo could you investigate?
> 
> At worst we might need to wrap the update script with `makeBinPath` or something.

Figured it out - see https://github.com/NixOS/nixpkgs/pull/332229/commits/a71c2d6d7bf2ae3aa73202870908e36def870e25

🙂

**Comment by DontEatOreo** [comment]
Sorry for the delay! I will take a look rn.

**Comment by jnsgruk** [comment]
> Sorry for the delay! I will take a look rn.

Don't worry - I think we're all sorted! Just waiting for the eval check and I'll merge.

**Comment by jnsgruk** [comment]
Result of `nixpkgs-review pr 332229` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>google-chrome</li>
    <li>netflix</li>
  </ul>
</details>

**Review by jnsgruk — APPROVED** [review]

**Comment by DontEatOreo** [comment]
LGTM!

image

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 332229` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>google-chrome</li>
    <li>netflix</li>
  </ul>
</details>

**Review by JohnRTitor — APPROVED** [review]
I squashed the small commits, should be good to go.

**Comment by JohnRTitor** [comment]
This needs a manual backport.

**Pull request merged**
